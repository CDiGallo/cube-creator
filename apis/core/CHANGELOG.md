# @cube-creator/core-api

## 0.2.2

### Patch Changes

- d96a3da: Transformation: pipeline would fail on fetching shared dimensions
- 6af426f: Output the default value for empty CSV cells
- fb600a6: Managed dimensions: Core API to save mappings with dimension metadata
- 8c82ec5: Do not show error when there is no cube (yet)
- Updated dependencies [bc22b24]
- Updated dependencies [fb600a6]
- Updated dependencies [876d7c0]
- Updated dependencies [3a25d90]
  - express-rdf-request@0.1.0
  - hydra-box-middleware-shacl@1.0.0
  - @cube-creator/managed-dimensions-api@0.1.0

## 0.2.1

### Patch Changes

- 0035c7d: Observations not loading on Cube Designer

## 0.2.0

### Minor Changes

- 47619a1: Vocabulary: use https://cube.link/ vocabulary (#510)

### Patch Changes

- 3574501: Metadata: use `dct:creator` for opendata.swiss publisher
- 77aa964: Redefine supported languages and their order (#393)
- 02a5b51: Metadata: added Data Kind field to dimension metadata (#490)
- 3574501: Metadata: fixed the `CreativeWorkStatus` terms (wrong case)
- 8ef5ec4: Metadata: Gram and kilogram were not available as units to select
- 3574501: Metadata: changed `dcterms:publisher` to simple text field
- Updated dependencies [47619a1]
- Updated dependencies [77aa964]
  - @cube-creator/core@0.2.0
  - @cube-creator/model@0.1.16
