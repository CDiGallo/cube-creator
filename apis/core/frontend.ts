import fs from 'fs'
import { Request, Response } from 'express'
import env from './lib/env'

export function uiConfig(req: Request, res: Response): void {
  res.header('content-type', 'application/javascript')

  if (env.has('AUTH_CONFIG_FILE')) {
    const stream = fs.createReadStream(env.AUTH_CONFIG_FILE)
    stream.pipe(res)
  } else {
    res.write(`
window.oidc = {
  authority: '${env.AUTH_ISSUER}',
  clientId: '${env.AUTH_CLIENT_ID}',
  scope: 'openid email profile pipelines:read pipelines:write',
}`)
  }
  res.end()
}
