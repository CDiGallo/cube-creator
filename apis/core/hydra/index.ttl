@prefix hydra: <http://www.w3.org/ns/hydra/core#> .
@prefix code:  <https://code.described.at/> .

<api>
  a
    hydra:ApiDocumentation ;
  hydra:entrypoint
    <> ;
  hydra:supportedClass
    hydra:Resource,
    hydra:Collection .

hydra:Resource
  hydra:supportedOperation
    [
      hydra:method
        "GET" ;
      code:implementedBy
        [
          a
            code:EcmaScript ;
          code:link
            <node:@hydrofoil/labyrinth/resource#get>
        ]
    ] .

hydra:Collection
  hydra:supportedOperation
    [
      hydra:method
        "GET" ;
      code:implementedBy
        [
          a
            code:EcmaScript ;
          code:link
            <node:@hydrofoil/labyrinth/collection#get>
        ]
    ] .
