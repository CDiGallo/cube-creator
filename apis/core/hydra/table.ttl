BASE          <urn:hydra-box:api>
PREFIX cc:    <https://cube-creator.zazuko.com/vocab#>
prefix hydra: <http://www.w3.org/ns/hydra/core#>
prefix code:  <https://code.described.at/>
prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

cc:Table
  a
    hydra:Class ;
  hydra:supportedProperty
    [
      a
        hydra:SupportedProperty ;
      hydra:property
        cc:csvw ;
    ] ;
.

cc:csvw
  a
    rdf:Property,
    hydra:Link ;
  hydra:supportedOperation
    [
      a
        hydra:Operation ;
      hydra:title
        "Get Table's CSVW mappings" ;
      hydra:method
        "GET" ;
      code:implementedBy
        [
          a
            code:EcmaScript ;
          code:link
            <file:handlers/table/csvw#get> ;
        ]
    ]
.
