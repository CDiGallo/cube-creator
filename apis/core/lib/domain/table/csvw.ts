import * as Csvw from '@rdfine/csvw'
import { TableMixin as CsvwTable } from '@rdfine/csvw'
import { CsvSourceMixin, TableMixin } from '@cube-creator/model'
import RdfResource from '@tpluscode/rdfine'
import cf from 'clownface'
import $rdf from 'rdf-ext'
import { csvw, xsd } from '@tpluscode/rdf-ns-builders'
import { NamedNode } from 'rdf-js'
import { ResourceStore } from '../../ResourceStore'
import { resourceStore } from '../resources'

RdfResource.factory.addMixin(...Object.values(Csvw))

interface Command {
  tableResource: NamedNode
  resources?: ResourceStore
}

export async function createCsvw({
  tableResource,
  resources = resourceStore(),
}: Command): Promise<Csvw.Table> {
  const table = new TableMixin.Class(await resources!.get(tableResource))
  const source = new CsvSourceMixin.Class(await resources!.get(table.csvSource.id))

  const {
    dialect: {
      quoteChar, delimiter, encoding, doubleQuote,
    },
  } = source

  return new CsvwTable.Class(cf({ dataset: $rdf.dataset(), term: table.csvw }), {
    url: source.id.value,
    dialect: {
      types: [csvw.Dialect],
      quoteChar,
      delimiter,
      encoding,
      doubleQuote,
    },
    tableSchema: {
      types: [csvw.Schema],
      aboutUrl: 'https://environment.ld.admin.ch/foen/ubd/28/{pollutant_id}/observation/{station_id}-{year}-{aggregation_id}',
      column: [{
        types: [csvw.Column],
        title: 'unit_id',
        datatype: xsd.string,
        propertyUrl: 'https://environment.ld.admin.ch/foen/ubd/28/unit-id',
      }, {
        types: [csvw.Column],
        title: 'aggregation_id',
        valueUrl: 'https://environment.ld.admin.ch/foen/ubd/28/aggregation/{aggregation_id}',
        propertyUrl: 'https://environment.ld.admin.ch/foen/ubd/28/aggregation',
      }, {
        types: [csvw.Column],
        title: 'station_id',
        valueUrl: 'https://environment.ld.admin.ch/foen/ubd/28/station/{station_id}',
        propertyUrl: 'https://environment.ld.admin.ch/foen/ubd/28/station',
      }, {
        types: [csvw.Column],
        title: 'value',
        datatype: xsd.float,
        propertyUrl: 'https://environment.ld.admin.ch/foen/ubd/28/dimension/value',
      }, {
        types: [csvw.Column],
        title: 'year',
        datatype: xsd.gYear,
        propertyUrl: 'https://environment.ld.admin.ch/foen/ubd/28/dimension/year',
      }, {
        types: [csvw.Column],
        suppressOutput: true,
        title: 'pollutant_id',
      }, {
        types: [csvw.Column],
        suppressOutput: true,
        title: 'limitvalue',
      }, {
        types: [csvw.Column],
        suppressOutput: true,
        title: 'value_remark',
      }],
    },
  })
}
