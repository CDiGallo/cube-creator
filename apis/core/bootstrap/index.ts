import * as core from '@cube-creator/core/bootstrap'
import env from '@cube-creator/core/env'
import { log } from '../lib/log'
import { cubeProjects } from './cube-projects'
import { entrypoint } from './entrypoint'
import { organizations } from './organizations'
import shapes from './shapes'

const resources = [cubeProjects, entrypoint, organizations, ...shapes]

export async function bootstrap(storeUrl: string, base: string) {
  log('Bootstrapping API resources')
  await core.bootstrap({
    storeUrl,
    base,
    user: env.maybe.STORE_ENDPOINTS_USERNAME,
    password: env.maybe.STORE_ENDPOINTS_PASSWORD,
    resources,
  })
}
