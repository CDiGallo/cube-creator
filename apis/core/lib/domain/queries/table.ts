import { NamedNode, Term } from 'rdf-js'
import { SELECT } from '@tpluscode/sparql-builder'

import { parsingClient } from '../../query-client'
import { cc } from '@cube-creator/core/namespace'

export async function * getTablesForMapping(csvMapping: NamedNode, client = parsingClient) {
  const results = await SELECT
    .DISTINCT`?table`
    .WHERE`
      GRAPH ?table
      {
        ?table a ${cc.Table} ;
             ${cc.csvMapping} ${csvMapping} .
      }
      `
    .execute(client.query)

  for (const result of results) {
    const table = result.table
    if (table.termType === 'NamedNode') {
      yield table
    }
  }
}

export async function * getLinkedTablesForSource(csvSource: NamedNode, client = parsingClient) {
  const results = await SELECT
    .DISTINCT`?table`
    .WHERE`
      GRAPH ?table
      {
        ?table a ${cc.Table} ;
             ${cc.csvSource} ${csvSource} .
      }
      `
    .execute(client.query)

  for (const result of results) {
    const table = result.table
    if (table.termType === 'NamedNode') {
      yield table
    }
  }
}

export async function getTableForColumnMapping(columnMapping: NamedNode, client = parsingClient): Promise<Term> {
  const results = await SELECT
    .DISTINCT`?table`
    .WHERE`
    GRAPH ?table
    {
      ?table a ${cc.Table} ;
           ${cc.columnMapping} ${columnMapping} .
    }
    `
    .execute(client.query)

  if (results.length < 1) {
    throw new Error(`No table for column mapping ${columnMapping.value} found`)
  }
  if (results.length > 1) {
    throw new Error(`More than one table for column mapping ${columnMapping.value} found`)
  }
  return results[0].table
}
