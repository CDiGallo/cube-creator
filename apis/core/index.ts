import express from 'express'
import { error, log } from './lib/log'
import authentication from './lib/auth'
import env from './lib/env'
import { uiConfig } from './frontend'

import guard = require('express-jwt-permissions')

async function main() {
  log('Starting Core API. Environment %s', env.production ? 'production' : 'development')

  const app = express()

  app.enable('trust proxy')

  app.get('/', (req, res) => {
    return res.status(200).end()
  })

  app.get('/env-config.js', uiConfig)

  app.use(authentication())

  app.get('/protected', guard().check('scope'), (req, res) => {
    return res.status(200).end()
  })

  app.listen(45670, () => log('Api ready'))
}

main().catch(error)
