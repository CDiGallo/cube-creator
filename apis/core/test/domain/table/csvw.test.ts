import { describe, it } from 'mocha'
import { expect } from 'chai'
import $rdf from 'rdf-ext'
import clownface from 'clownface'
import { cc } from '@cube-creator/core/namespace'
import { csvw, schema } from '@tpluscode/rdf-ns-builders'
import { createCsvw } from '../../../lib/domain/table/csvw'
import { TestResourceStore } from '../../support/TestResourceStore'

describe('domain/table/csvw', () => {
  describe('createCsvw', () => {
    it('copies csvw:dialect from CsvSource', async function () {
      // given
      const tableResource = $rdf.namedNode('cc:table')
      const sourceId = $rdf.namedNode('cc:source')
      const resources = new TestResourceStore([
        clownface({ dataset: $rdf.dataset() })
          .node(tableResource)
          .addOut(cc.csvSource, sourceId)
          .addOut(cc.csvw, $rdf.namedNode('cc:table/csvw')),
        clownface({ dataset: $rdf.dataset() })
          .node(sourceId)
          .addOut(schema.associatedMedia, media => {
            media.addOut(schema.contentUrl, $rdf.namedNode('content url'))
          })
          .addOut(csvw.dialect, dialect => {
            dialect.addOut(csvw.quoteChar, '""')
            dialect.addOut(csvw.separator, '-')
          }),
      ])

      // when
      const table = await createCsvw({
        resources,
        tableResource,
      })

      // then
      expect(table.dialect?.toJSON()).to.matchSnapshot(this)
    })

    it('copies sources contentUrl to csvw:url', async function () {
      // given
      const tableResource = $rdf.namedNode('cc:table')
      const sourceId = $rdf.namedNode('cc:source')
      const resources = new TestResourceStore([
        clownface({ dataset: $rdf.dataset() })
          .node(tableResource)
          .addOut(cc.csvSource, sourceId)
          .addOut(cc.csvw, $rdf.namedNode('cc:table/csvw')),
        clownface({ dataset: $rdf.dataset() })
          .node(sourceId)
          .addOut(schema.associatedMedia, media => {
            media.addOut(schema.contentUrl, $rdf.namedNode('content url'))
          })
          .addOut(csvw.dialect),
      ])

      // when
      const table = await createCsvw({
        resources,
        tableResource,
      })

      // then
      expect(table.url).to.equal('content url')
    })
  })
})
