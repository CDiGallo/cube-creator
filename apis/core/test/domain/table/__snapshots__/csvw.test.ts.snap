exports["domain/table/csvw createCsvw copies csvw:dialect from CsvSource"] = {
	"id": "_:b3",
	"type": [
		"http://www.w3.org/ns/csvw#Dialect"
	],
	"quoteChar": "\"\"",
	"@context": {
		"id": "@id",
		"type": "@type",
		"quoteChar": "http://www.w3.org/ns/csvw#quoteChar"
	}
};

